packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts

```diff
@@ -43,14 +60,108 @@ export default class MicrosoftAuth {
         'email',
         'User.Read',
       ],
+      scopeTransform = scopes => scopes.concat('offline_access'),
     } = options;
 
-    return OAuth2.create({
-      discoveryApi,
-      oauthRequestApi,
-      provider,
-      environment,
-      defaultScopes,
-    });
+    this.environment = environment;
+    this.provider = provider;
+    this.oauthRequestApi = oauthRequestApi;
+    this.discoveryApi = discoveryApi;
+    this.scopeTransform = scopeTransform;
+
+    this.oauth2 = {
+      [MicrosoftAuth.MicrosoftGraphID]: OAuth2.create({
+        discoveryApi: this.discoveryApi,
+        oauthRequestApi: this.oauthRequestApi,
+        provider: this.provider,
+        environment: this.environment,
+        scopeTransform: this.scopeTransform,
+        defaultScopes,
+      }),
+    };
+  }
+
+  private microsoftGraph(): OAuth2 {
+    return this.oauth2[MicrosoftAuth.MicrosoftGraphID];
+  }
+
+  static expectedClaims(scope: string): { aud: string; scp: string } {
+    const scopes = scope.split(' ').map(MicrosoftAuth.parseScope);
+    const firstAudience =
```

Comment:
Rather than picking the first one and essentially leaving the rest broken, do you think it would make sense to be more strict and throw an error if there's a conflict in the requested scopes? I'm feeling that'll help users of this API figure out errors way earlier compared to having to figure out why their requested access token only has one of the two scopes requested. Frankly I feel this is something that it would make sense if the MS OAuth implementation would do ._.

That's if we can do this reliably and there's no valid case for passing conflicting scopes.

Reply:
b76cbe84cc54983ffe95b67eaa308d928d649e23